pkg/domain/extension.go

```diff
@@ -390,87 +361,705 @@ func (e *ErrExtensionCredentialsNotFound) Error() string {
+		service.EnsureIDs(ctx, e)
+	}
+}
+
+// AddService adds a new service to the extension.
+func (e *Extension) AddService(ctx context.Context, service *ExtensionService) (*ExtensionService, error) {
+	service.EnsureIDs(ctx, e)
+
+	if e.Services == nil {
+		e.Services = make(map[string]*ExtensionService)
+	} else {
+		if _, err := e.GetService(ctx, service.ID); err == nil {
+			return nil, NewErrExtensionServiceExists(e.ID, service.ID)
+		}
+	}
+
+	service.SetCreated(ctx, time.Now())
+	e.Services[service.ID] = service
+
+	return service, nil
+}
+
+// AddEndpoint adds a new endpoint to a service.
+func (e *Extension) AddEndpoint(ctx context.Context, serviceID string, endpoint *ExtensionServiceEndpoint) (*ExtensionServiceEndpoint, error) {
+	service, error := e.GetService(ctx, serviceID)
+	if error != nil {
+		return nil, error
+	}
+	return service.AddEndpoint(ctx, endpoint)
+}
+
+// AddCredential adds a new credential to a service.
+func (e *Extension) AddCredential(ctx context.Context, serviceID string, credential *ExtensionServiceCredential) (*ExtensionServiceCredential, error) {
+	service, error := e.GetService(ctx, serviceID)
+	if error != nil {
+		return nil, error
+	}
+	return service.AddCredential(ctx, credential)
+}
+
+// GetServices returns all services belonging to the extension.
+func (e *Extension) GetServices(ctx context.Context) []*ExtensionService {
+	if e.Services == nil {
+		return []*ExtensionService{}
+	}
+
+	services := make([]*ExtensionService, 0, len(e.Services))
+	for _, service := range e.Services {
+		services = append(services, service)
+	}
+
+	return services
+}
+
+// GetService returns a service belonging to the extension.
+func (e *Extension) GetService(ctx context.Context, serviceID string) (*ExtensionService, error) {
+	if e.Services != nil {
+		service, ok := e.Services[serviceID]
+		if ok {
+			return service, nil
+		}
+	}
+
+	return nil, NewErrExtensionServiceNotFound(e.ID, serviceID)
+}
+
+// GetServiceEndpoints returns all endpoints belonging to a service.
+func (e *Extension) GetServiceEndpoints(ctx context.Context, serviceID string) ([]*ExtensionServiceEndpoint, error) {
+	service, error := e.GetService(ctx, serviceID)
+	if error != nil {
+		return nil, error
+	}
+
+	return service.GetEndpoints(ctx), nil
+}
+
+// GetServiceEndpoint returns an endpoint belonging to a service.
+func (e *Extension) GetServiceEndpoint(ctx context.Context, serviceID string, endpointID string) (*ExtensionServiceEndpoint, error) {
+	service, err := e.GetService(ctx, serviceID)
+	if err != nil {
+		return nil, err
+	}
+
+	endpoint, err := service.GetEndpoint(ctx, endpointID)
+	if err != nil {
+		return nil, NewErrExtensionServiceEndpointNotFound(e.ID, serviceID, endpointID)
+	}
+
+	return endpoint, nil
+}
+
+// GetServiceCredentials returns all credentials belonging to a service.
+func (e *Extension) GetServiceCredentials(ctx context.Context, serviceID string) ([]*ExtensionServiceCredential, error) {
```

Comment:
I can see why it's difficult to keep "credentials" as the name of the structure. You could practice `GetAllServiceCredentials` or `GetServiceCredentialsList` for the entire list of objects and keep `GetServiceCredentials` for a single object. But you'd probably need to refactor the method names of all other objects as well, to be consistent (e.g. `GetAllServices` or `GetServiceList` instead of `GetServices`)